src/forwarddiff.jl

```diff
@@ -0,0 +1,51 @@
+using .ForwardDiff: ForwardDiff, DiffResults, StaticArrays
+
+struct ForwardDiffBackend <: AbstractBackend end
+
+@primitive function pushforward_function(ba::ForwardDiffBackend, f, xs...)
+    return function pushforward(vs)
+        return ForwardDiff.derivative(h -> f(step_toward.(xs, vs, h)...), 0)
+    end
+end
+function pushforward_function(::ForwardDiffBackend, f, x)
+    return function pushforward(v)
+        if v isa Tuple
+            @assert length(v) == 1
+            return (ForwardDiff.derivative(h -> f(step_toward(x, v[1], h)), 0),)
+        else
+            return (ForwardDiff.derivative(h -> f(step_toward(x, v, h)), 0),)
+        end
+    end
+end
+
+primal_value(x::ForwardDiff.Dual) = ForwardDiff.value(x)
+
+# these implementations are more efficient than the fallbacks
+
+gradient(::ForwardDiffBackend, f, x::AbstractArray) = (ForwardDiff.gradient(f, x),)
+
+function jacobian(ba::ForwardDiffBackend, f, x::AbstractArray)
+    y = f(x)
```

Comment:
is this function call necessary? can't we always use the `vect` approach?

Reply:
Good point. `Base.vect` doesn't for that, but we can do:
```julia
@inline _asarray(x) = [x]
@inline _asarray(x::AbstractArray) = x
```